src/FakeItEasy/Core/FakeManager.cs

```diff
@@ -143,14 +141,10 @@ void IFakeCallProcessor.Process(IInterceptedFakeObjectCall fakeObjectCall)
                 listener.OnBeforeCallIntercepted(fakeObjectCall);
             }
 
-            var ruleToUse =
-                (from rule in this.AllRules
-                 where rule.Rule.IsApplicableTo(fakeObjectCall) && rule.HasNotBeenCalledSpecifiedNumberOfTimes()
-                 select rule).First();
-
+            IFakeObjectCallRule ruleUsed = null;
             try
             {
-                ApplyRule(ruleToUse, fakeObjectCall);
+                this.ApplyBestRule(fakeObjectCall, out ruleUsed);

```

Comment:
Is there a reason to return `ruleUsed` as an out parameter here? Can't it just be returned directly, since there's no other return value?

Reply:
It was a way to set the `ruleUsed` (and ensure that the caller knew what it was) before attempting the application of the rule. If the `.Apply` throws, we won't get an opportunity to return the rule. I could wrap each `Apply` call in `ApplyBestRule` with try/catches and return the rule regardless.

Reply:
Ah, I see. You can leave it as is, then.

Reply:
Sorry, I was unclear. In progress, I'm replacing `ApplyBestRule` with 

```
private IFakeObjectCallRule ChooseBestRule(IInterceptedFakeObjectCall fakeObjectCall)
{
    foreach (var preUserRule in this.preUserRules)
    {
        if (preUserRule.IsApplicableTo(fakeObjectCall))
        {
            return preUserRule;
        }
    }

    foreach (var rule in this.AllUserRules)
    {
        if (rule.Rule.IsApplicableTo(fakeObjectCall) && rule.HasNotBeenCalledSpecifiedNumberOfTimes())
        {
            rule.CalledNumberOfTimes++;
            return rule.Rule;
        }
    }

    foreach (var postUserRule in PostUserRules)
    {
        if (postUserRule.IsApplicableTo(fakeObjectCall))
        {
            return postUserRule;
        }
    }

    // The last rule in the postUserRules is applicable to all calls, so this line is here to satisfy
    // the compiler. We will always have returned by now.
    return null;
}
```

And then we can just call `Apply` right after. I think it's better than the `out`. But if you prefer the `out`, I can leave.

Reply:
Ah, I see. What bothers me with this approach is that `CalledNumberOfTimes` is incremented in the method that chooses the rule, rather than in the method that applies it...

Reply:
A fair point. Back to `out` with an explanatory comment?

Reply:
Sure, a comment wouldn't hurt!